File: src/Dotnet9.Tools.Web/Models/BlogPost.cs
namespace Dotnet9.Tools.Web.Models;

public class BlogPost
{
    public string? Title { get; set; }

    public string? Slug { get; set; }

    public string? Description { get; set; }

    public string? Cover { get; set; }

    public string[]? Categories { get; set; }

    public string[]? Tags { get; set; }

    public string[]? Albums { get; set; }

    public string? CopyrightType { get; set; }

    public string? Original { get; set; }

    public string? OriginalLink { get; set; }

    public string? CreateDate { get; set; }

    public string? UpdateDate { get; set; }

    public string? Content { get; set; }
}

public static class BlogPostExtensions
{
    public static bool IsExist(this BlogPost post, string filter)
    {
        if (!string.IsNullOrWhiteSpace(post.Title) && post.Title.ToLower().Contains(filter.ToLower())) return true;

        if (!string.IsNullOrWhiteSpace(post.Slug) && post.Slug.ToLower().Contains(filter.ToLower())) return true;

        if (!string.IsNullOrWhiteSpace(post.Description) &&
            post.Description.ToLower().Contains(filter.ToLower())) return true;

        if (post.Categories != null && post.Categories.Contains(filter)) return true;

        if (post.Tags != null && post.Tags.Contains(filter)) return true;

        if (post.Albums != null && post.Albums.Contains(filter)) return true;

        if (!string.IsNullOrWhiteSpace(post.Original) && post.Original.ToLower().Contains(filter.ToLower()))
            return true;

        if (!string.IsNullOrWhiteSpace(post.OriginalLink) &&
            post.OriginalLink.ToLower().Contains(filter.ToLower())) return true;

        return !string.IsNullOrWhiteSpace(post.Content) && post.Content.ToLower().Contains(filter.ToLower());
    }
}

File: src/Dotnet9.Web.Go/service/blogpost.go
package service

import (
	"html/template"

	"dotnet9.com/goweb/config"
	"dotnet9.com/goweb/dao"
	"dotnet9.com/goweb/models"
	"github.com/russross/blackfriday"
)

func GetBlogPostDetail(slug string) (*models.BlogPostRes, error) {
	blogpost, err := dao.GetBlogPostBySlug(slug)
	
	if err != nil {
		return nil, err
	}
	categoriesOfBlogPost := dao.GetCagegoriesByBlogPostId(blogpost.Id)
	contentBytes := []byte(blogpost.Markdown)
	contentStr := string(blackfriday.MarkdownCommon(contentBytes))
	blogPostMore := models.BlogPostMore{
		Id:                   blogpost.Id,
		Title:                blogpost.Title,
		Slug:                 blogpost.Slug,
		Description:          blogpost.Description,
		Cover:                blogpost.Cover,
		Markdown: 		   	  blogpost.Markdown,
		Content: 			  template.HTML(contentStr),
		CopyrightType:        blogpost.CopyrightType,
		OriginalAvatar:       blogpost.OriginalAvatar,
		OriginalTitle:        blogpost.OriginalTitle,
		OriginalLink:         blogpost.OriginalLink,
		Categories:           categoriesOfBlogPost,
		ViewCount:            blogpost.ViewCount,
		CreationTime:         models.DateDay(blogpost.CreationTime),
		LastModificationTime: models.DateDay(blogpost.LastModificationTime),
	}
	if len(blogpost.Original) == 0 {
		blogPostMore.Original = config.AppCfg.Site.DefaultAuthor
	} else {
		blogPostMore.Original = blogpost.Original
	}
	var postRes = &models.BlogPostRes{
		Viewer: config.AppCfg.Viewer,
		SystemConfig: config.AppCfg.System,
		Article: blogPostMore,
	}
	return postRes, nil
}